### PR TITLE
Configurable colors for wireframe

### PR DESCRIPTION
**Pull request by IceSentry**
# Objective

- Make the wireframe colors configurable at the global level and the single mesh level
- Based on https://github.com/bevyengine/bevy/pull/5314

This video shows what happens when playing with various settings from the example

https://github.com/bevyengine/bevy/assets/8348954/1ee9aee0-fab7-4da8-bc5d-8d0562bb34e6

## Solution

- Add a `color` field to the `WireframeMaterial`
- Use a `WireframeColor` component to configure the color per entity
- Add a `default_color` field to `WireframeConfig` for global wireframes or wireframes with no specified color.

## Notes

- Most of the docs and the general idea for `WireframeColor` came from [UberLambda](https://github.com/UberLambda) in #3677 but the code ended up completely different so I created a separate branch. ~~I'm not sure how to correctly credit them on this PR.~~ (I re-created the commit but I added them as co-author in the commit message)

~~Closes https://github.com/bevyengine/bevy/pull/3677~~
~~Closes https://github.com/bevyengine/bevy/pull/5301~~

~~https://github.com/bevyengine/bevy/pull/5314 should be merged before this PR.~~

### CODE REVIEW

**Comment by superdump** [comment]
Could you restructure the code to follow the order of the original so that the diff is cleaner? It’s a bit all over the place removing and inserting lines that are from practically the same code in different places in the file which makes it difficult to see what actually changed and so, difficult to review.

**Comment by IceSentry** [comment]
Done. It's still a bit noisy because the approach is so different, but it's much better now and that also made me realize I removed the Reflect stuff by mistake.

**Comment by IceSentry** [comment]
@superdump should I make a PR with only the update to the material system and make another PR that adds per-entity colors after that? It would probably make it easier to review, but it felt like the PR was small enough to review all in one go. It would help make the diffs less noisy though.

File: crates/bevy_pbr/src/wireframe.rs
Comment (alice-i-cecile) [inline crates/bevy_pbr/src/wireframe.rs:50]: We should explain the behavior if this component is absent, but [`Wireframe`] is missing.

File: crates/bevy_pbr/src/wireframe.rs
Comment (alice-i-cecile) [inline crates/bevy_pbr/src/wireframe.rs:43]: This should link to `WireframeColor`.

File: crates/bevy_pbr/src/wireframe.rs
Comment (alice-i-cecile) [inline crates/bevy_pbr/src/wireframe.rs:177]: Should this be pub? Seems like something others may want to reuse.

File: examples/3d/wireframe.rs
Comment (JMS55) [inline examples/3d/wireframe.rs:12]: I assume that this needs to be inserted before DefaultPlugins is added? We should note that here if so

File: crates/bevy_pbr/src/wireframe.rs
Comment (JMS55) [inline crates/bevy_pbr/src/wireframe.rs:109]: I think this is supposed to say without a WireframeColor component?

**Review by JMS55 — COMMENTED** [review]
I'm a bit confused on the different components. Can you explain the structure of how things work? Why is WireframeColor a separate component, instead of an optional field of Wireframe?

**Comment by IceSentry** [comment]
I somewhat forgot about this PR, I'll try to update it.

For the WireframeColor component, I don't remember the detaisl, but I believe this was discussed in this PR https://github.com/bevyengine/bevy/pull/3677

Also, I originally wanted us to merge, https://github.com/bevyengine/bevy/pull/5314 first and then add color support just to make the PRs easier to review and merge.

**Comment by IceSentry** [comment]
Right, so the WireframeColor component makes it way easier to use change detection to detect when/if the color changed

File: examples/3d/wireframe.rs
Comment (IceSentry) [reply examples/3d/wireframe.rs:12]: This is fixed now with the current plugin group design

**Review by alice-i-cecile — APPROVED** [review]

**Comment by nerdachse** [comment]
This is approved but will not come with 0.12? :(

**Comment by alice-i-cecile** [comment]
@nerdachse this needs a second approval, from any community member, before I can merge it (hint). It's not in the milestone (so we won't block on it), but if it's merged in time it'll be shipped for 0.12.

**Review by nerdachse — APPROVED** [review]
I don't know about the finished conversation https://github.com/bevyengine/bevy/pull/5303/files?diff=split&w=0#r1025198770

but apart from that, I reviewed it and I approve the changes.

I am not sure about potential performance pitfalls or such, but the code changes look clean and understandable.

**Comment by IceSentry** [comment]
@nerdachse the main blocker is waiting on https://github.com/bevyengine/bevy/pull/5314 the issue is that most rendering devs are currently working on other bigger reviews but once those are done both wireframe PRs should be merged fairly quickly.

**Comment by alice-i-cecile** [comment]
@IceSentry the prerequisite PR has been merged now; can you clean up merge conflicts when you get a chance?

**Comment by alice-i-cecile** [comment]
Thanks :) FYI, I'm going to wait to merge this until https://github.com/bevyengine/bevy/pull/9258 is through: Rob's asked me to hold off on merging rendering PRs until that's done to avoid generating complex conflicts for the author.

**Comment by IceSentry** [comment]
I don't think there would be any conflict with this PR, but I'm in no hurry to merge this PR either.

**Comment by nerdachse** [comment]
If deferred rendering for some reason doesn't make it into 0.12 it would still be awesome to have this.

I am highly anticipating this feature and would hate to wit another 3 months.

Also, while I understand that there are PRs that warrant holding off other work, I argue that they should be sparse and not hold too much other valuable work "hostage" so ti speak.

Just my 0.05 of course. I am very grateful for both, deferred rendering and this, actually most that's happening in bevy! So please don't take this the wrong way!

**Comment by alice-i-cecile** [comment]
Yep, this will be in 0.12 one way or another :) Definitely agree that blocking work to get specific PRs through should be very rare: I think the only other time we've done that this cycle is for Assets v2.

File: crates/bevy_pbr/src/wireframe.rs
Comment (JMS55) [inline crates/bevy_pbr/src/wireframe.rs:120]: https://dev-docs.bevyengine.org/bevy/ecs/prelude/fn.resource_changed.html

**Review by JMS55 — CHANGES_REQUESTED** [review]

File: crates/bevy_pbr/src/wireframe.rs
Comment (alice-i-cecile) [reply crates/bevy_pbr/src/wireframe.rs:120]: Good choice: run conditions will be marginally more performant (and better reflect intent).

**Review by JMS55 — APPROVED** [review]

**Pull request merged**
